### ISSUE TITLE
[data grid] Format row count numbers in pagination

### ISSUE THREAD

**Issue by Janpot** [mui/mui-x#13222]
### Summary

Live demo https://stackblitz.com/edit/react-jxqb4n?file=Demo.tsx

The pagination control looks like:

![Screenshot 2024-05-23 at 14 58 46](https://github.com/mui/mui-x/assets/2109932/28ba69db-78de-49ac-8b9a-8442cf5536c4)

Would it be possible to format these numbers? A thousands separator would make them more readable. In practice this is mostly going to be interesting for the total count, but once you start formatting, you may as well do it for all of these numbers.

### Examples

![Screenshot 2024-05-23 at 15 05 47](https://github.com/mui/mui-x/assets/2109932/ce36af1b-1adb-4418-a4d4-7f0102ae1f4c)

**Search keywords**: pagination format number

**Comment by michelengelen** [comment]
Probably already possible with a [custom pagination](https://mui.com/x/react-data-grid/components/#pagination) and the [TablePagination](https://mui.com/material-ui/api/table-pagination/) component ... the prop `labelDisplayedRows` looks like it would do the trick here

But this seems like a very heavy customization for such a little thing!

@mui/xgrid WDYT?

**Comment by flaviendelangle** [comment]
Looking at the code, it seems that you have a `labelDisplayedRows` locale key on the `MuiTablePagination` locale entry of the grid (those nested locale entry are super hard to understand IMHO).

This would be nice to have an example if it does work
And if it doesn't and you do need to create a custom pagination component, then I fully agree that we could improve the DX :+1:

**Comment by michelengelen** [comment]
Oh yes, that actually works!

<img width="394" alt="Screenshot 2024-05-23 at 16 36 50" src="https://github.com/mui/mui-x/assets/32863416/bcc497a4-5ed5-4401-a307-a332038aaa8b">

FYI: numbers were artificially created to test this (`(10000).toLocaleString('en')`) 

Thanks @flaviendelangle for this tip! 👍🏼 

```tsx
slotProps={{
  pagination: { labelDisplayedRows: ({ from, to, count }) => `${from.toLocaleString('en')}-${to.toLocaleString('en')}`},
}}
```

**Comment by flaviendelangle** [comment]
@michelengelen do you think it would justify a doc example?

**Comment by Janpot** [comment]
Respecting the decision ofcourse but gonna play a bit advocate of the devil here. To me it seems strange that the `localeText` is not formatting the numbers it interpollates by default. Imagining there was a text with a date, we wouldn't just stringify that date neither, we'd format it to the user locale. It feels a bit like this API is only doing half its job this way. I find it hard to imagine that not formatting the numbers would be part of more than 5% of use-cases out there.

**Comment by flaviendelangle** [comment]
Related issues #7457

**Comment by flaviendelangle** [comment]
Yeah, I forget that one every time :laughing: 
The Base UI team is experiencing similar issues with the `NumberInput`, maybe we will be able to use the same approach in the end.

**Comment by Janpot** [comment]
Actually, if the text is already localized, why can't the localization of the numbers follow the text? If there is an SSR problem, shouldn't it then be already present in the current localization of the texts?

**Comment by oliviertassinari** [comment]
Related to https://github.com/mui/material-ui/issues/24495

**Comment by siriwatknp** [comment]
Here is my proposal:

- add a util to format the pagination (only when it's number)
- apply the util to all of the locales so that the result matches each locale

See https://github.com/mui/mui-x/compare/master...siriwatknp:material-ui-x:fix/format-pagination

---

Or an alternative is to update only the docs (with the sample code) to show how to add commas to pagination

**Comment by MBilalShafi** [comment]
> Here is my proposal:
> 
>     * add a util to format the pagination (only when it's number)
> 
>     * apply the util to all of the locales so that the result matches each locale
> 
> 
> See [siriwatknp@d13ee7f](https://github.com/siriwatknp/material-ui-x/commit/d13ee7f08e9f13c7e2ff7a8192bee540e13f40a5)

What should be the other locales' strategy in this case? Should `formatNumber()` also accept `Intl.NumberFormat` constructor args (or at least locale name) to honor [locale based formatting](https://github.com/mui/mui-x/issues/13222#issuecomment-2131368232)?

I'd also expect some users experiencing this change as breaking.

WDYT about starting with a docs example and targeting this one for v9 BCs?

**Comment by siriwatknp** [comment]
> What should be the other locales' strategy in this case? Should `formatNumber()` also accept `Intl.NumberFormat` constructor args (or at least locale name) to honor [locale based formatting](https://github.com/mui/mui-x/issues/13222#issuecomment-2131368232)?

Yes, I pushed the comment too early. The new version is https://github.com/mui/mui-x/compare/master...siriwatknp:material-ui-x:fix/format-pagination with locale `csCZ` update as an example.

> WDYT about starting with a docs example and targeting this one for v9 BCs?

Absolutely, docs is a good starting point as I mentioned an alternative. I will draft the docs PR and wait for feedback about my proposal above (it can be v9 but let's decide if it's a good approach or not).

**Comment by MBilalShafi** [comment]
> Yes, I pushed the comment too early. The new version is https://github.com/mui/mui-x/compare/master...siriwatknp:material-ui-x:fix/format-pagination with locale csCZ update as an example.

Looks good to me 👍 
It should match most of the users' expectations and the rest could use `labelDisplayedRows` locale key override.
I'd be in favor of adding it to the [v9 breaking changes](https://github.com/mui/mui-x/issues/17846).

**Comment by Janpot** [comment]
This is great. Just a quick review of the code but we could bundle optimize a little bit:
```tsx
  const numValue = typeof value === 'string' ? Number(value) : value;

  if (Number.isFinite(numValue) && typeof Intl !== 'undefined' && Intl.NumberFormat) {
    try {
      return new Intl.NumberFormat(locale).format(numValue);
    } catch {
      // fall through
    }
  }

  return String(numValue);
```

Should we allow passing options to `.format()`? I mean we may want to e.g. format percentages or so at some point.

e.g. in bundle size dashboard we create multiple [formatters](https://github.com/mui/mui-public/blob/57c5c9205302acc01ce4508087e26d9b1e6345e1/apps/code-infra-dashboard/src/components/SizeChangeDisplay.tsx#L5-L43).

**Comment by oliviertassinari** [comment]
How about we fix https://github.com/mui/material-ui/issues/24495 first? This problem seems nothing data grid related, so this seems to be:

1. An issue to fix in Material UI
2. Then to propagate in MUI X Data Grid (e.g. aggregation, pagination, pivoting), MUI X Charts (e.g. tooltip, axis), etc.

**Comment by siriwatknp** [comment]
@oliviertassinari sure, they are independent. The approach I am proposing is to update the locale configuration, since both projects offer the configuration, they are not dependent.

- Material UI: https://github.com/mui/material-ui/blob/master/packages/mui-material/src/locale/index.ts
- DataGrid: https://github.com/siriwatknp/material-ui-x/tree/docs/theme-tokens-visual/packages/x-data-grid/src/locales

**Comment by oliviertassinari** [comment]
> update the locale configuration

@siriwatknp It's not clear if this is the best strategy.

In https://github.com/mui/material-ui/issues/24495#issue-788973679, it was also about exploring having a global provider that can propagate to all the locales, without having to recreate them. This comes at the cost of having to support a runtime number formating API, which is heavier than your prototype in https://github.com/mui/mui-x/issues/13222#issuecomment-3183062461.

But yeah, it's possible that dating the locale configuration might be good enough, the most pragmatic, that maybe most people don't need to customize the number formater (if they do, they would need to redo all the localization strings impacted, in all the locales they support).

**Comment by siriwatknp** [comment]
> But yeah, it's possible that dating the locale configuration might be good enough, the most pragmatic

Agree.

> if they do, they would need to redo all the localization strings impacted, in all the locales they support

That would be rare case which is acceptable. I bet that user would go with formatted by default.

Next step, apply [my proposal](https://github.com/mui/mui-x/issues/13222#issuecomment-3183062461) in the next major release
